ipld/src/selector/walk.rs

```diff
@@ -0,0 +1,203 @@
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        Progress {
+            resolver,
+            path: Path::default(),
+        }
+        .walk_all(ipld, self, &callback)
+        .await
+    }
+
+    /// Walks a graph of Ipld nodes, executing the callback only on the nodes "matched".
+    /// If a resolver is passed in, links will be able to be traversed.
+    pub async fn walk_matching<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
+        callback: F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        self.walk_all(ipld, resolver, |prog, ipld, reason| -> Result<(), String> {
+            if let VisitReason::SelectionMatch = reason {
+                return callback(prog, ipld);
+            }
+            Ok(())
+        })
+        .await
+    }
+}
+
+/// Provides reason for callback in traversal for `walk_all`.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum VisitReason {
+    /// Ipld node visited was a specific match.
+    SelectionMatch,
+    /// Ipld node was visited while searching for matches.
+    SelectionCandidate,
+}
+
+#[async_trait]
+pub trait LinkResolver {
+    #[allow(unused_variables)]
+    /// Resolves a Cid link into it's respective Ipld node, if it exists.
+    async fn load_link(&self, link: &Cid) -> Result<Option<Ipld>, String> {
+        Err("load_link not implemented on the LinkResolver".into())
+    }
+}
+
+impl LinkResolver for () {}
+
+#[derive(Clone, Debug)]
+pub struct Progress<L = ()>
+where
+    L: Clone,
+{
+    resolver: Option<L>,
+    path: Path,
+}
+
+impl<L> Progress<L>
+where
+    L: LinkResolver + Sync + Send + Clone,
+{
+    /// Returns the path of the current progress
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    #[async_recursion]
+    async fn walk_all<F>(
+        &mut self,
+        ipld: &Ipld,
+        selector: Selector,
+        callback: &F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld, VisitReason) -> Result<(), String> + Sync,
+    {
+        let reason = if selector.decide() {
+            VisitReason::SelectionMatch
+        } else {
+            VisitReason::SelectionCandidate
+        };
+        callback(self, ipld, reason).map_err(Error::Custom)?;
+
+        // If Ipld is list or map, continue traversal, otherwise return
+        match ipld {
+            Ipld::Map(_) | Ipld::List(_) => (),
+            _ => return Ok(()),
+        }
+
+        match selector.interests() {
+            Some(interests) => {
+                for ps in interests {
+                    let v = match ipld.lookup_segment(&ps) {
+                        Some(ipld) => ipld,
+                        None => continue,
+                    };
+                    self.clone()
+                        .traverse_node(ipld, selector.clone(), callback, ps, v)
+                        .await?
```

Comment:
can we use a filter_map + map here? interests.filter_map.(...).map(..)

Reply:
I'm a bit confused by this, can you explain? Do you mean filtering based on the available paths of the ipld?

Reply:
filtering based on the ipld.look_up_segment with a filtermap essentially. would save us from having the continue clause there since filter_map is based on options. I guess the issue though is that you might have to unwrap v in the map clause but it should be already filtered out

Reply:
yeah I had considered, I think the reason I hadn't was because integer path segments can be used ambiguously as fields of a map, but let me revisit to remember why I didn't go through with it

Reply:
The trouble here is that calling `filter_map` with `ipld.lookup_segment` will give you an iterator of `Ipld`s without the paths (which you still need) so it does indeed become pretty verbose:

```rust
for (ps, ipld) in interests
    .into_iter()
    .filter_map(|ps| ipld.lookup_segment(&ps).map(|ipld| (ps, ipld)))
{
    // ...
}
```
Not really an improvement. Unless there's a way to improve this which I haven't thought of